File: app/src/main/java/com/example/taxconnect/adapter/ServiceAdapter.java
package com.example.taxconnect.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.example.taxconnect.databinding.ItemServiceBinding;
import com.example.taxconnect.model.ServiceModel;
import java.util.ArrayList;
import java.util.List;

public class ServiceAdapter extends RecyclerView.Adapter<ServiceAdapter.ServiceViewHolder> {

    private List<ServiceModel> services = new ArrayList<>();
    private final OnServiceClickListener listener;
    private OnServiceDeleteListener deleteListener;
    private boolean isEditable = false;

    public interface OnServiceClickListener {
        void onServiceClick(ServiceModel service);
    }

    public interface OnServiceDeleteListener {
        void onServiceDelete(ServiceModel service);
    }

    public ServiceAdapter(OnServiceClickListener listener) {
        this.listener = listener;
    }

    public void setOnServiceDeleteListener(OnServiceDeleteListener listener) {
        this.deleteListener = listener;
    }

    public void setEditable(boolean editable) {
        this.isEditable = editable;
        notifyDataSetChanged();
    }

    public void setServices(List<ServiceModel> services) {
        this.services = services;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ServiceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemServiceBinding binding = ItemServiceBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ServiceViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ServiceViewHolder holder, int position) {
        holder.bind(services.get(position));
    }

    @Override
    public int getItemCount() {
        return services.size();
    }

    class ServiceViewHolder extends RecyclerView.ViewHolder {
        private final ItemServiceBinding binding;

        public ServiceViewHolder(ItemServiceBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(ServiceModel service) {
            binding.tvTitle.setText(service.getTitle());
            binding.tvDesc.setText(service.getDescription());
            binding.tvPrice.setText("\u20B9 " + service.getPrice());
            binding.tvTime.setText("Time: " + service.getEstimatedTime());
            
            if (isEditable) {
                binding.btnBuy.setText("Edit");
                binding.btnDelete.setVisibility(android.view.View.VISIBLE);
            } else {
                binding.btnBuy.setText("Buy Now");
                binding.btnDelete.setVisibility(android.view.View.GONE);
            }

            binding.btnBuy.setOnClickListener(v -> {
                if (listener != null) listener.onServiceClick(service);
            });

            binding.btnDelete.setOnClickListener(v -> {
                if (deleteListener != null) deleteListener.onServiceDelete(service);
            });
        }
    }
}

File: app/src/main/java/com/example/taxconnect/features/ca/ServiceAdapter.kt
package com.example.taxconnect.features.ca

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.taxconnect.R
import com.example.taxconnect.data.models.ServiceModel

class ServiceAdapter(private val onServiceClick: (ServiceModel) -> Unit) :
    RecyclerView.Adapter<ServiceAdapter.ServiceViewHolder>() {

    private var services: List<ServiceModel> = ArrayList()
    private var isEditable: Boolean = false
    private var onDeleteListener: ((ServiceModel) -> Unit)? = null

    fun setServices(services: List<ServiceModel>) {
        this.services = services
        notifyDataSetChanged()
    }

    fun setEditable(editable: Boolean) {
        this.isEditable = editable
        notifyDataSetChanged()
    }

    fun setOnServiceDeleteListener(listener: (ServiceModel) -> Unit) {
        this.onDeleteListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ServiceViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_service, parent, false)
        return ServiceViewHolder(view)
    }

    override fun onBindViewHolder(holder: ServiceViewHolder, position: Int) {
        val service = services[position]
        holder.bind(service, isEditable, onServiceClick, onDeleteListener)
    }

    override fun getItemCount(): Int = services.size

    class ServiceViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvTitle: TextView = itemView.findViewById(R.id.tvTitle)
        private val tvPrice: TextView = itemView.findViewById(R.id.tvPrice)
        private val tvDesc: TextView = itemView.findViewById(R.id.tvDesc)
        private val tvTime: TextView = itemView.findViewById(R.id.tvTime)
        private val btnDelete: ImageButton = itemView.findViewById(R.id.btnDelete)
        private val btnBuy: Button = itemView.findViewById(R.id.btnBuy)

        fun bind(
            service: ServiceModel,
            isEditable: Boolean,
            onServiceClick: (ServiceModel) -> Unit,
            onDeleteListener: ((ServiceModel) -> Unit)?
        ) {
            tvTitle.text = service.title
            tvPrice.text = "₹ ${service.price}"
            tvDesc.text = service.description
            tvTime.text = "Time: ${service.estimatedTime}"

            if (isEditable) {
                btnDelete.visibility = View.VISIBLE
                btnBuy.visibility = View.GONE
                btnDelete.setOnClickListener { onDeleteListener?.invoke(service) }
                itemView.setOnClickListener { onServiceClick(service) }
            } else {
                btnDelete.visibility = View.GONE
                btnBuy.visibility = View.VISIBLE
                btnBuy.setOnClickListener { onServiceClick(service) }
                itemView.setOnClickListener(null)
            }
        }
    }
}
